spec/abstract/AARCH64/ArchVSpace_A.thy

```diff
@@ -0,0 +1,686 @@
+definition invalidate_vmid_entry :: "vmid \<Rightarrow> (unit,'z::state_ext) s_monad" where
+  "invalidate_vmid_entry vmid \<equiv> do
+     vmid_table \<leftarrow> gets (arm_vmid_table \<circ> arch_state);
+     vmid_table' \<leftarrow> return (vmid_table (vmid := None));
+     modify (\<lambda>s. s \<lparr> arch_state := (arch_state s) \<lparr> arm_vmid_table := vmid_table' \<rparr>\<rparr>)
+  od"
+
+text \<open>Remove mappings in either direction involving this ASID.\<close>
+definition invalidate_asid_entry :: "asid \<Rightarrow> (unit,'z::state_ext) s_monad" where
+  "invalidate_asid_entry asid \<equiv> do
+     maybe_hw_asid \<leftarrow> load_vmid asid;
+     when (maybe_hw_asid \<noteq> None) $ invalidate_vmid_entry (the maybe_hw_asid);
+     invalidate_asid asid
+  od"
+
+text \<open>Locate a VMID that is not in use, if necessary by reclaiming one already assigned to an ASID.\<close>
+definition find_free_vmid :: "(vmid,'z::state_ext) s_monad" where
+  "find_free_vmid \<equiv> do
+     vmid_table \<leftarrow> gets (arm_vmid_table \<circ> arch_state);
+     next_vmid \<leftarrow> gets (arm_next_vmid \<circ> arch_state);
+     maybe_vmid \<leftarrow> return $ find (\<lambda>a. vmid_table a = None)
+                                 (take (length [minBound :: vmid .e. maxBound])
+                                       ([next_vmid .e. maxBound] @ [minBound .e. next_vmid]));
+     case maybe_vmid of
+       Some vmid \<Rightarrow> return vmid
+     | None \<Rightarrow> do
+         invalidate_asid $ the $ vmid_table next_vmid;
+         do_machine_op $ invalidateTranslationASID (ucast next_vmid);
+         invalidate_vmid_entry next_vmid;
+         modify (\<lambda>s. s \<lparr> arch_state := (arch_state s) \<lparr> arm_next_vmid := next_vmid + 1 \<rparr>\<rparr>);
+         return next_vmid
+       od
+   od"
+
+text \<open>Get the VMID associated with an ASID, assigning one if none is already assigned.\<close>
+definition get_vmid :: "asid \<Rightarrow> (vmid, 'z::state_ext) s_monad" where
+  "get_vmid asid \<equiv> do
+     maybe_vmid \<leftarrow> load_vmid asid;
+     case maybe_vmid of
+       Some vmid \<Rightarrow> return vmid
+     | None \<Rightarrow>  do
+         new_hw_asid \<leftarrow> find_free_vmid;
+         store_vmid asid new_hw_asid;
+         return new_hw_asid
+       od
+   od"
+
+text \<open>
+  Format a VM fault message to be passed to a thread's supervisor after it encounters a page fault.\<close>
+definition handle_vm_fault :: "obj_ref \<Rightarrow> vmfault_type \<Rightarrow> (unit, 'z::state_ext) f_monad" where
+  "handle_vm_fault thread fault \<equiv> case fault of
+     ARMDataAbort \<Rightarrow> doE
+       addr \<leftarrow> liftE $ do_machine_op getFAR;
+       fault \<leftarrow> liftE $ do_machine_op getDFSR;
+       cur_v \<leftarrow> liftE $ gets (arm_current_vcpu \<circ> arch_state);
+       addr \<leftarrow> if (\<exists>v. cur_v = Some (v, True)) \<comment> \<open>VCPU active\<close>
```

Comment:
The Haskell has `curVCPUActive`, but it is only called twice in total. Since it's a read-only function I suspect it's easier to keep it unfolded in this particular case, instead of having to state a bunch of lemmas about it. I might be wrong, but we can still tweak this when the proofs get annoying.